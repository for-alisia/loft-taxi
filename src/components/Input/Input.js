import React from 'react';
import PropTypes, { shape } from 'prop-types';
import './Input.scss';

const Input = ({
    onInputChange,
    label,
    hasError,
    errorMsg,
    validation,
    classes,
    ...otherProps
}) => {
    const inputId = `input${Math.random()}`;
    const className = ['form-input'];
    const isInvalid = hasError ? true : false;

    if (hasError) {
        className.push('invalid');
    }

    if (classes) {
        className.push(classes);
    }

    return (
        <div className='form-input-wrapper'>
            <input
                className={className.join(' ')}
                id={inputId}
                onChange={onInputChange}
                {...otherProps}
            />
            <label className='form-label' htmlFor={inputId}>
                {label}
                {validation ? <span>*</span> : null}
            </label>
            {isInvalid ? (
                <span className='form-input-error'>{errorMsg}</span>
            ) : null}
        </div>
    );
};

Input.propTypes = {
    onInputChange: PropTypes.func,
    label: PropTypes.string,
    hasError: PropTypes.bool,
    errorMsg: PropTypes.string,
    classes: PropTypes.string,
    otherProps: shape({
        type: PropTypes.string.isRequired,
        placeholder: PropTypes.string,
        name: PropTypes.string
    })
};

export default Input;
