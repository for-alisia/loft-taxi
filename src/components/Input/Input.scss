@import '../../scss/abstracts.scss';
.form-input-wrapper {
    margin-bottom: rem(40);

    position: relative;
}

.form-input {
    display: block;
    width: 100%;

    padding: rem(5) 0;

    font-size: rem(16);
    line-height: rem(25);

    border: none;
    outline: none;
    border-bottom: 2px solid $color-grey;

    &:focus {
        outline: none;
        border-bottom: 2px solid $color-blue;
    }

    &.invalid {
        border-color: $color-red;
    }

    &.invalid + .form-label {
        color: $color-red;
    }
}

.form-label {
    display: block;
    width: 100%;

    font-size: rem(12);
    color: $color-grey-dark;

    transition: all 0.2s ease;
    transform: translateY(-3rem);
    opacity: 1;
}

.form-input:placeholder-shown + .form-label {
    opacity: 0;
    visibility: hidden;
    transform: translateY(0rem);
}

.form-input-error {
    position: absolute;

    color: $color-red;
    font-size: rem(12);

    bottom: -0.5rem;
    left: 0;
}
