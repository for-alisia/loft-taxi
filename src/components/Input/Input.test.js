import React from 'react';
import Input from './Input';
import { shallow } from 'enzyme';

describe('Input component', () => {
    const commonClass = '.form-input';

    it('renders input component', () => {
        expect(shallow(<Input />)).toMatchSnapshot();
    });

    it('renders invalid input', () => {
        const wrapper = shallow(<Input hasError={true} />);

        expect(wrapper.find(commonClass).hasClass('invalid')).toEqual(true);
    });

    it('calls function on each change', () => {
        const mockHandleChangeInput = jest.fn();
        const wrapper = shallow(
            <Input onInputChange={mockHandleChangeInput} />
        );

        wrapper.find(commonClass).simulate('change');
        wrapper.find(commonClass).simulate('change');

        expect(mockHandleChangeInput.mock.calls.length).toBe(2);
    });

    it('shows error block with invalid input', () => {
        const wrapper = shallow(<Input hasError={true} errorMsg='error' />);

        expect(wrapper.find('.form-input-error')).toHaveLength(1);
    });

    it('shows error message with invalid input', () => {
        const wrapper = shallow(<Input hasError={true} errorMsg='error' />);

        expect(wrapper.find('.form-input-error').text()).toEqual('error');
    });

    it('call handler after click event on icon', () => {
        const mockOnClearInput = jest.fn();
        const wrapper = shallow(<Input onClearInput={mockOnClearInput} />);

        wrapper.find('.form-input-icon').simulate('click');

        expect(mockOnClearInput.mock.calls.length).toBe(1);
    });

    it('shows * with required inputs', () => {
        const wrapper = shallow(<Input validation={{}} />);

        expect(wrapper.find('.form-label span')).toHaveLength(1);
    });

    it('contains correct type of input', () => {
        const wrapper = shallow(<Input type='password' />);

        expect(wrapper.find('[type="password"]')).toHaveLength(1);
    });
});
