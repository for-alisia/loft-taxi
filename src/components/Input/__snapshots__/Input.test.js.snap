// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input component renders input component 1`] = `ShallowWrapper {}`;
